packages/web-extension/src/walletManager/SignerManager/SignerManager.ts

```diff
@@ -0,0 +1,168 @@
+  SignTransactionProps,
+  SignerManagerConfirmationApi,
+  SignerManagerSignApi,
+  TransactionWitnessRequest
+} from './types';
+import { Subject } from 'rxjs';
+
+export type HardwareKeyAgentOptions = TrezorConfig;
+
+export type SignerManagerProps = {
+  hwOptions: HardwareKeyAgentOptions;
+};
+
+export type SignerManagerDependencies = {
+  keyAgentFactory: KeyAgentFactory;
+};
+
+const clearPassphrase = (passphrase: Uint8Array) => {
+  for (let i = 0; i < passphrase.length; i++) {
+    passphrase[i] = 0;
+  }
+};
+
+export class SignerManager<WalletMetadata extends {}>
+  implements SignerManagerConfirmationApi<WalletMetadata>, SignerManagerSignApi<WalletMetadata>
+{
+  readonly transactionWitnessRequest$ = new Subject<TransactionWitnessRequest<WalletMetadata>>();
+  readonly signDataRequest$ = new Subject<SignDataRequest<WalletMetadata>>();
+  readonly #hwOptions: HardwareKeyAgentOptions;
+  readonly #keyAgentFactory: KeyAgentFactory;
+
+  constructor(props: SignerManagerProps, { keyAgentFactory }: SignerManagerDependencies) {
+    this.#hwOptions = props.hwOptions;
+    this.#keyAgentFactory = keyAgentFactory;
+  }
+
+  async signTransaction(
+    { tx, signContext, options }: SignTransactionProps,
+    requestContext: RequestContext<WalletMetadata>
+  ): Promise<Cardano.Signatures> {
+    const transaction = Serialization.Transaction.fromCbor(tx);
+    return this.#signRequest(
+      this.transactionWitnessRequest$,
+      {
+        requestContext,
+        signContext,
+        transaction,
+        walletType: requestContext.wallet.type
+      },
+      (keyAgent) =>
+        keyAgent.signTransaction(
+          {
+            body: transaction.body().toCore(),
+            hash: transaction.getId()
+          },
+          signContext,
+          options
+        )
+    );
+  }
+
+  async signData(props: SignDataProps, requestContext: RequestContext<WalletMetadata>): Promise<SignBlobResult> {
+    return this.#signRequest(
+      this.signDataRequest$,
+      {
+        ...props,
+        requestContext,
+        walletType: requestContext.wallet.type
+      },
+      (keyAgent) => keyAgent.signBlob(props.derivationPath, props.blob)
+    );
+  }
+
+  #signRequest<R, Req extends RequestBase<WalletMetadata> & SignRequest<R>>(
+    emitter$: Subject<Req>,
+    request: Omit<Req, 'reject' | 'sign'>,
+    sign: (keyAgent: KeyAgent) => Promise<R>
+  ) {
+    return new Promise<R>((resolve, reject) => {
+      if (!emitter$.observed) {
+        return reject(new errors.AuthenticationError('Internal error: signDataRequest$ not observed'));
+      }
+      const account = request.requestContext.wallet.accounts.find(
+        ({ accountIndex }) => accountIndex === request.requestContext.accountIndex
+      );
+      if (!account) {
+        return reject(new errors.ProofGenerationError(`Account not found: ${request.requestContext.accountIndex}`));
+      }
+      const bubbleResolveReject = async (action: () => Promise<R>): Promise<R> => {
```

Comment:
This could be its own generic helper, not just within this context. 
And the #signRequest could in general be broken down into smaller maybe testable pieces.

Reply:
Hoisted bubbleResolveReject to module scope https://github.com/input-output-hk/cardano-js-sdk/pull/1019/commits/4c9498178292af741c97f027f85bbadd88241a73

> And the #signRequest could in general be broken down into smaller maybe testable pieces.

Maybe, but I don't see an obvious way to split it up. Let me know if you have some concrete ideas

Reply:
I agree it's not obvious, It just seems long, If I get an idea will let you know